### DISCUSSION TITLE
SentryEvent class does not contain a property that stores the scoped ServiceProvider

### DISCUSSION THREAD

**Discussion by LucasRomankiu** [getsentry/sentry-dotnet#893] [Q&A]
Please mark the type framework used:

- [ ] ASP.NET MVC
- [ ] ASP.NET Web API (OWIN)
- [X ] ASP.NET Core
- [ ] WPF
- [ ] WinForms
- [ ] Xamarin
- [ ] Other:

Please mark the type of the runtime used:

- [ ] .NET Framework
- [ ] Mono
- [X ] .NET Core

* Version: 3.1

Please mark the NuGet packages used:

- [ ] Sentry
- [ ] Sentry.Serilog
- [ ] Sentry.NLog
- [ ] Sentry.Log4Net
- [ ] Sentry.Extensions.Logging
- [X ] Sentry.AspNetCore

* Version: 2.1.6

Hello, i am trying to use the Sentry.AspNetCore for logging exceptions. But right now i'm facing a issue, i'm trying to put informations in my tags by modyfing the BeforeSend property and the problem is that the SentryEvent class does not contain a property for the scoped serviceProvider that contains the informations that i needed in the first place. 
`                            options.BeforeSend = sentryEvent =>
                            {
                                sentryEvent.ServiceProvider;
                                return sentryEvent;
                            };
`
I believe that we are able to get the service provider by changing the InvokeAsync method inside the SentryMiddleware class with the following change:
```
                void CaptureException(Exception e)
                {
                    var evt = new SentryEvent(e);

                    // ServiceProvider is the new property to be implemented that is responsible for storing the IServiceProvider
                    evt.ServiceProvider = context.RequestServices;

                    _logger?.LogTrace("Sending event '{SentryEvent}' to Sentry.", evt);

                    var id = hub.CaptureEvent(evt);

                    _logger?.LogInformation("Event '{id}' queued.", id);
                }
```
Is this a valid change? Is there another way of doing it?

**Comment by bruno-garcia** [comment]
Consider creating an EventProcessor instead and registering it in the DI container.
It allows you to modify or drop the event just like beforeSend but you can take stuff via constructor injection.

Here's an example registering it:
https://github.com/getsentry/sentry-dotnet/blob/15d1d7677f9bbbbe7ec2f307cc4a07b916d50e8e/samples/Sentry.Samples.AspNetCore.Mvc/Startup.cs#L25

here's an example processor:

https://github.com/getsentry/sentry-dotnet/blob/15d1d7677f9bbbbe7ec2f307cc4a07b916d50e8e/samples/Sentry.Samples.AspNetCore.Mvc/ExampleEventProcessor.cs#L7-L19

**Comment by LucasRomankiu** [comment]
I tried using the processors, and it works really well. But i'm having some problems related to using scoped services inside the 
 event processor. From what i saw Sentry is trying to get my event processors from ServiceProviderRoot, and is failing to get the scoped services for this reason. We're using scoped services due to the fact that we do not want to expose a variable with the serviceProvider, this would violate our design choices. The following is a HangFire code that executes backgroundJobs using a serviceProvider scope, the following is a real case using Hangfire where do not have the AsyncLocal variable to expose:
```
        [BackgroundJobQueueFilter]
        [AutomaticRetry(Attempts = 3)]
        public async Task Execute(BaseBackgroundJobInfo args, IJobCancellationToken jobCancellationToken = null)
        {
            using (var scope = _serviceProvider.CreateScope())
            {
                var context = GetBackgroundJobExecutionContext(scope.ServiceProvider, args);
                await ResolveAmbientData(scope.ServiceProvider);
                await SetMultiTenancyProperties(scope.ServiceProvider);
                await MediateStartupTasksAsync(scope.ServiceProvider);
                await _backgroundJobExecutor.ExecuteAsync(context, jobCancellationToken);
            }
        }
```
This means that i won't be able to inject scoped dependencies inside my ISentryEventProcessor, and since this is a backgroundJob there's no httpContext available. Is there any way to circumvent/fix this?

**Comment by bruno-garcia** [comment]
This seems related to #318 

In your case there might be a work around, something like:

```csharp
using Sentry;

[BackgroundJobQueueFilter]
[AutomaticRetry(Attempts = 3)]
public async Task Execute(BaseBackgroundJobInfo args, IJobCancellationToken jobCancellationToken = null)
{
    using (var scope = _serviceProvider.CreateScope())
    {
        var hub = scope.Resolve<IHub>()
        using (hub.PushScope())
        {
            hub.ConfigureScope(s => {
              s.AddEventProcessor(scope.Resolve<WhateverScopedDependency>());
            });
            // Now if events are captured below this point, the processor added above will be executed
            var context = GetBackgroundJobExecutionContext(scope.ServiceProvider, args);
            await ResolveAmbientData(scope.ServiceProvider);
            await SetMultiTenancyProperties(scope.ServiceProvider);
            await MediateStartupTasksAsync(scope.ServiceProvider);
            await _backgroundJobExecutor.ExecuteAsync(context, jobCancellationToken);
        }
    }
}
```

I didn't run this but I hope it explains the work around I'm suggesting. It will create a Sentry scope, you can add the event processor you want in that scope, so while inside the `using` block, that processor will run for all events that go through Sentry's SDK.

And since you have the correct scoped service provider, it should work as you expect.

The solution to #318 still would need to be worked on. Maybe we need to do what I suggested above inside the `SentryMiddleware`, we already push a scope, but we could then bind the local serviceLocator to that scope instead of relying on the ones added at the start of the app (root service locator).
